src/java/org/apache/cassandra/io/util/File.java

```diff
@@ -0,0 +1,608 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.io.util;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.channels.FileChannel;
+import java.nio.file.*;
+import java.util.Objects;
+import java.util.function.BiPredicate;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import javax.annotation.Nullable;
+
+import com.google.common.util.concurrent.RateLimiter;
+
+import org.apache.cassandra.io.FSWriteError;
+
+import static org.apache.cassandra.io.util.PathUtils.filename;
+import static org.apache.cassandra.utils.Throwables.maybeFail;
+
+/**
+ * A thin wrapper around java.nio.file.Path to provide more ergonomic functionality.
+ *
+ * TODO codebase probably should not use tryList, as unexpected exceptions are hidden;
+ *      probably want to introduce e.g. listIfExists
+ * TODO codebase probably should not use Paths.get() to ensure we can override the filesystem
+ */
+public class File implements Comparable<File>
+{
+    private static final FileSystem filesystem = FileSystems.getDefault();
+    public enum WriteMode { OVERWRITE, APPEND }
+
+    public static String pathSeparator()
+    {
+        return filesystem.getSeparator();
+    }
+
+    @Nullable final Path path; // nullable to support concept of empty path, that resolves to the working directory if converted to an absolute path
+
+    /**
+     * Construct a File representing the child {@code child} of {@code parent}
+     */
+    public File(String parent, String child)
+    {
+        this(parent.isEmpty() ? null : filesystem.getPath(parent), child);
+    }
+
+    /**
+     * Construct a File representing the child {@code child} of {@code parent}
+     */
+    public File(File parent, String child)
+    {
+        this(parent.path, child);
+    }
+
+    /**
+     * Construct a File representing the child {@code child} of {@code parent}
+     */
+    public File(Path parent, String child)
+    {
+        // if "empty abstract path" (a la java.io.File) is provided, we should behave as though resolving relative path
+        if (child.startsWith(pathSeparator()))
+            child = child.substring(pathSeparator().length());
+        this.path = parent == null ? filesystem.getPath(child) : parent.resolve(child);
+    }
+
+    /**
+     * Construct a File representing the provided {@code path}
+     */
+    public File(String path)
+    {
+        this(path.isEmpty() ? null : filesystem.getPath(path));
```

Comment:
Not required for this patch, just sharing some ideas. Different `path` may have different file systems, eg. commitlog maybe stored locally but sstable will be stored remotely; Or each keyspace may have its own FS.

Reply:
We could definitely support multiple file systems, but to do so across the codebase would require phasing out the use of this `String` constructor, since the paths supplied to `java.io.File` are plain paths referring to the default file system. I would definitely support this in a future patch, and hopefully this work has made that more manageable.